Address FIXME for ASSERT(!m_buffer) in VectorBuffer::allocateBuffer

abigailfox opened the pull request:
#### 892a15725852530aee0c064a88e7703345b80af6
<pre>
Address FIXME for ASSERT(!m_buffer) in VectorBuffer::allocateBuffer

<a href="https://bugs.webkit.org/show_bug.cgi?id=250801">https://bugs.webkit.org/show_bug.cgi?id=250801</a>
rdar://104401340

Reviewed by NOBODY (OOPS!).

Address a FIXME comment in WTF::Vector to ensure that a new buffer cannot be allocated if a buffer already exists for that VectorBuffer
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/892a15725852530aee0c064a88e7703345b80af6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105548 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14638 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38445 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114800 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174944 "Failed to checkout and rebase branch from PR 9072") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16058 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5861 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97843 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114634 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111304 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12215 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95195 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39703 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94079 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26831 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81405 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/95308 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7924 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28187 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/93448 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/5698 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8290 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4774 "Found 60 new test failures: accessibility/ARIA-reflection.html, accessibility/accessibility-crash-focused-element-change.html, accessibility/accessibility-crash-setattribute.html, accessibility/accessibility-crash-with-dynamic-inline-content.html, accessibility/accessibility-node-reparent.html, accessibility/custom-elements/autocomplete.html, accessibility/custom-elements/controls-shadow.html, accessibility/custom-elements/controls.html, accessibility/custom-elements/current.html, accessibility/custom-elements/describedby-shadow.html ... (failure)") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30388 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14048 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47730 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/102158 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9944 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25471 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WTF/wtf/Vector.h:
```diff
@@ -496,6 +496,7 @@ class VectorBuffer : private VectorBufferBase<T, Malloc> {
     bool allocateBuffer(size_t newCapacity)
     {
         // FIXME: This should ASSERT(!m_buffer) to catch misuse/leaks. https://bugs.webkit.org/show_bug.cgi?id=250801
+        ASSERT(!m_buffer);
```
Please remove the comment now that you fixed it.  (The comment serves no purpose after adding the assertion.)

ddkilzer replied:
Actually, this is hitting a lot of failures on [mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4006), so those need to be investigated first.

ddkilzer replied:
Abigail and I took a look at the code in `Vector.h`, and we found that one of the following two conditions must be true where the `ASSERT()` was added:

1. `m_buffer` is `nullptr`, or
2. `m_buffer` is set to `inlineBuffer()`.

Because condition #2 wasn't being checked in the original patch, that caused crashes on Debug builds.

ddkilzer reviewed (approved):
Going to approve this, but please remove the comment before landing.

ddkilzer reviewed (changes requested):
Need to fix Debug assertions on Debug bot before landing!

Either the `ASSERT()` condition needs another check, or maybe we're re-allocating a buffer unnecessarily and this caught bug(s).

abigailfox committed:
Address FIXME for ASSERT(!m_buffer) in VectorBuffer::allocateBuffer

https://bugs.webkit.org/show_bug.cgi?id=250801
rdar://104401340

Reviewed by NOBODY (OOPS!).

Address a FIXME comment in WTF::Vector to ensure that a new buffer cannot be allocated if a buffer already exists for that VectorBuffer
